File: wagerproof-ios-native/Wagerproof/Features/Agents/Components/AgentLeaderboard.swift
import SwiftUI
import WagerproofDesign
import WagerproofModels
import WagerproofStores

/// Native port of `components/agents/AgentLeaderboard.tsx`. Renders:
///   - The filter pill rows (sort mode + timeframe).
///   - A list of `LeaderboardRow`s — top 3 wrapped in a glow halo.
///   - `EmptyState` when zero public agents.
///   - Skeletons during the initial fetch.
///
/// Owns no state of its own; the parent injects a `LeaderboardStore` and the
/// view binds directly to its `sortMode` / `timeframe` properties (their
/// didSet triggers a refresh).
struct AgentLeaderboardView: View {
    @Bindable var store: LeaderboardStore
    @Environment(AgentEntitlementsStore.self) private var entitlements
    /// When false the sort/timeframe pills are hidden — the host (Outliers tab)
    /// lifts `LeaderboardFilterBar` into its pinned glass header instead.
    var showsFilters: Bool = true
    var onRowTap: (AgentLeaderboardEntry) -> Void

    var body: some View {
        ScrollView {
            LazyVStack(alignment: .leading, spacing: 12) {
                if showsFilters {
                    LeaderboardFilterBar(store: store)
                }
                content
            }
            .padding(.horizontal, Spacing.lg)
            .padding(.top, Spacing.md)
            .padding(.bottom, Spacing.xl)
        }
        .refreshable {
            await store.refresh()
        }
        .task {
            if case .idle = store.loadState {
                await store.refresh()
            }
        }
    }

    // MARK: - Content branches

    @ViewBuilder
    private var content: some View {
        switch store.loadState {
        case .idle, .loading:
            if store.entries.isEmpty {
                ForEach(0..<6, id: \.self) { _ in skeletonRow }
            } else {
                rows
            }
        case .loaded:
            if store.entries.isEmpty {
                emptyState
            } else {
                rows
            }
        case .failed(let msg):
            ContentUnavailableView {
                Label("Couldn't load leaderboard", systemImage: "exclamationmark.triangle")
            } description: {
                Text(msg)
            } actions: {
                Button {
                    Task { await store.refresh() }
                } label: {
                    Label("Retry", systemImage: "arrow.clockwise")
                }
                .buttonStyle(.borderedProminent)
                .tint(Color(hex: 0x00E676))
            }
            .padding(.top, Spacing.xl)
        }
    }

    private var rows: some View {
        ForEach(Array(store.entries.enumerated()), id: \.element.id) { idx, entry in
            let rank = idx + 1
            LeaderboardRow(
                entry: entry,
                rank: rank,
                lockStats: !entitlements.canViewAgentPicks,
                isEntitlementsLoading: entitlements.isLoading,
                isBottomMode: store.isBottomMode,
                onTap: { onRowTap(entry) }
            )
            // Cascade each row in when it replaces the loading shimmer.
            .staggeredAppear(index: idx)
        }
    }

    private var emptyState: some View {
        VStack(spacing: 10) {
            Image(systemName: "trophy")
                .font(.system(size: 48))
                .foregroundStyle(Color.appTextSecondary)
            Text("No public agents yet")
                .font(.system(size: 16, weight: .semibold))
                .foregroundStyle(Color.appTextSecondary)
            Text("Be the first to make your agent public!")
                .font(.system(size: 13))
                .foregroundStyle(Color.appTextSecondary)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, Spacing.xxl)
        .background(
            RoundedRectangle(cornerRadius: 12, style: .continuous)
                .fill(Color.appBorder.opacity(0.2))
        )
    }

    /// Skeleton placeholder for `LeaderboardRow`. Mirrors that row's layout
    /// (rank block, avatar circle, name + sport subtitle, trailing record/units
    /// + win-rate badge) using the unified Skeleton* primitives, inside the same
    /// 12pt chrome. The inner group carries the `.shimmering()` sweep; the card
    /// chrome stays solid (applied after the shimmer).
    private var skeletonRow: some View {
        HStack(spacing: 12) {
            SkeletonBlock(width: 28, height: 20, cornerRadius: 4)
            SkeletonCircle(36)
            VStack(alignment: .leading, spacing: 2) {
                SkeletonBlock(width: 120, height: 14)
                SkeletonBlock(width: 60, height: 10)
            }
            Spacer(minLength: 0)
            VStack(alignment: .trailing, spacing: 2) {
                SkeletonBlock(width: 44, height: 12)
                SkeletonBlock(width: 52, height: 14)
            }
            SkeletonBlock(width: 40, height: 11)
        }
        .padding(12)
        .shimmering()
        .background(
            RoundedRectangle(cornerRadius: 12, style: .continuous).fill(Color.appBorder.opacity(0.2))
        )
    }
}

// MARK: - LeaderboardFilterBar

/// The leaderboard's sort-mode + timeframe pill rows (and the 10+ picks
/// toggle), extracted so it can render either inline in `AgentLeaderboardView`
/// or lifted into the Outliers tab's pinned glass header. Binds directly to the
/// shared `LeaderboardStore` — each pill tap flips a property whose didSet
/// re-runs the fetch.
struct LeaderboardFilterBar: View {
    @Bindable var store: LeaderboardStore

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(LeaderboardStore.SortMode.allCases, id: \.self) { mode in
                        filterPill(label: mode.label, isActive: store.sortMode == mode) {
                            store.sortMode = mode
                        }
                    }
                }
                .padding(.vertical, 2)
            }

            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(LeaderboardStore.Timeframe.allCases, id: \.self) { tf in
                        filterPill(label: tf.label, isActive: store.timeframe == tf, isSubtle: true) {
                            store.timeframe = tf
                        }
                    }
                    Toggle("10+ picks", isOn: $store.excludeUnder10Picks)
                        .toggleStyle(.button)
                        .controlSize(.small)
                        .tint(Color(hex: 0x00E676))
                }
                .padding(.vertical, 2)
            }
        }
    }

    private func filterPill(label: String, isActive: Bool, isSubtle: Bool = false, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Text(label)
                .font(.system(size: 12, weight: .bold))
                .padding(.horizontal, 12)
                .padding(.vertical, 6)
                .background(
                    Capsule().fill(
                        isActive
                            ? Color(hex: 0x00E676).opacity(0.15)
                            : Color.appBorder.opacity(isSubtle ? 0.3 : 0.5)
                    )
                )
                .overlay(
                    Capsule().stroke(
                        isActive
                            ? Color(hex: 0x00E676).opacity(0.45)
                            : Color.appBorder.opacity(0.3),
                        lineWidth: 1
                    )
                )
                .foregroundStyle(isActive ? Color(hex: 0x00E676) : Color.appTextSecondary)
        }
        .buttonStyle(.plain)
        .sensoryFeedback(.selection, trigger: isActive)
    }
}

// MARK: - LeaderboardRow

/// One row of `AgentLeaderboardView`. Mirrors RN `LeaderboardRow`.
private struct LeaderboardRow: View {
    let entry: AgentLeaderboardEntry
    let rank: Int
    let lockStats: Bool
    let isEntitlementsLoading: Bool
    let isBottomMode: Bool
    let onTap: () -> Void

    var body: some View {
        Button(action: onTap) {
            HStack(spacing: 12) {
                rankBadge
                avatarSection
                Spacer(minLength: 0)
                statsSection
                winRateBadge
                Image(systemName: "chevron.right")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Color.appTextSecondary)
            }
            .padding(12)
            .background(
                RoundedRectangle(cornerRadius: 12, style: .continuous)
                    .fill(Color.appBorder.opacity(0.2))
            )
            .overlay(
                RoundedRectangle(cornerRadius: 12, style: .continuous)
                    .strokeBorder(Color.appBorder.opacity(0.3), lineWidth: 1)
            )
        }
        .buttonStyle(.plain)
    }

    @ViewBuilder
    private var rankBadge: some View {
        let (color, icon): (Color, String?) = {
            switch rank {
            case 1: return (Color(hex: 0xFFD700), "trophy.fill")
            case 2: return (Color(hex: 0xC0C0C0), "medal.fill")
            case 3: return (Color(hex: 0xCD7F32), "medal")
            default: return (Color.appTextSecondary, nil)
            }
        }()

        if let icon {
            Image(systemName: icon)
                .font(.system(size: 22, weight: .semibold))
                .foregroundStyle(color)
                .frame(width: 32)
        } else {
            Text("\(rank)")
                .font(.system(size: 16, weight: .bold))
                .foregroundStyle(color)
                .frame(width: 32)
        }
    }

    @ViewBuilder
    private var avatarSection: some View {
        let avatar = ZStack {
            Circle().fill(LinearGradient(
                colors: AgentColorPalette.avatarGradient(for: entry.avatarColor),
                startPoint: .topLeading, endPoint: .bottomTrailing))
            PixelSpriteAvatar(spriteIndex: entry.spriteIndex)
                .padding(rank <= 3 ? 3 : 2)
        }
        .frame(width: rank <= 3 ? 44 : 36, height: rank <= 3 ? 44 : 36)

        HStack(spacing: 10) {
            if rank <= 3 {
                GlowingCardWrapper(color: entry.avatarColor, cornerRadius: 22) {
                    avatar
                }
            } else {
                avatar
            }
            VStack(alignment: .leading, spacing: 2) {
                Text(entry.name)
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(Color.appTextPrimary)
                    .lineLimit(1)
                HStack(spacing: 4) {
                    ForEach(Array(entry.preferredSports.prefix(2)), id: \.self) { sport in
                        Text(sport.label)
                            .font(.system(size: 10, weight: .semibold))
                            .foregroundStyle(Color.appTextSecondary)
                    }
                    if entry.preferredSports.count > 2 {
                        Text("+\(entry.preferredSports.count - 2)")
                            .font(.system(size: 10, weight: .semibold))
                            .foregroundStyle(Color.appTextSecondary)
                    }
                }
            }
        }
    }

    private var statsSection: some View {
        VStack(alignment: .trailing, spacing: 2) {
            Text(record)
                .font(.system(size: 12, weight: .bold))
                .foregroundStyle(Color.appTextPrimary)
            ZStack {
                if isEntitlementsLoading {
                    ProgressView().controlSize(.mini)
                } else {
                    Text(netUnitsLabel)
                        .font(.system(size: 14, weight: .heavy))
                        .foregroundStyle(entry.netUnits >= 0 ? Color.appWin : Color.appLoss)
                }
                if !isEntitlementsLoading && lockStats {
                    // Lock overlay — simulates the AndroidBlurView blur on RN.
                    ZStack {
                        Rectangle()
                            .fill(.ultraThinMaterial)
                        Image(systemName: "lock.fill")
                            .font(.system(size: 9))
                            .foregroundStyle(Color.appTextSecondary)
                    }
                }
            }
        }
        .frame(minWidth: 56)
    }

    private var winRateBadge: some View {
        let label = entry.winRate.map { String(format: "%.1f%%", $0 * 100) } ?? "-"
        let color: Color = {
            if isBottomMode {
                if let wr = entry.winRate, wr < 0.35 { return .appLoss }
                return Color(hex: 0xF97316)
            }
            return Color(hex: 0x00E676)
        }()
        return Text(label)
            .font(.system(size: 11, weight: .bold))
            .foregroundStyle(color)
            .frame(width: 48)
    }

    private var record: String {
        var s = "\(entry.wins)-\(entry.losses)"
        if entry.pushes > 0 { s += "-\(entry.pushes)" }
        return s
    }

    private var netUnitsLabel: String {
        let sign = entry.netUnits >= 0 ? "+" : ""
        return String(format: "%@%.2fu", sign, entry.netUnits)
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/agents/components/AgentLeaderboard.kt
package com.wagerproof.app.features.agents.components

import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.composed
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch
import com.wagerproof.core.design.components.SkeletonBlock
import com.wagerproof.core.design.components.SkeletonCircle
import com.wagerproof.core.design.components.shimmering
import com.wagerproof.core.design.components.staggeredAppear
import com.wagerproof.core.design.pixeloffice.PixelSpriteAvatar
import com.wagerproof.core.design.tokens.AppColors
import com.wagerproof.core.models.AgentLeaderboardEntry
import com.wagerproof.core.services.AgentPerformanceService
import com.wagerproof.core.stores.AgentEntitlementsStore
import com.wagerproof.core.stores.LeaderboardStore
import com.wagerproof.core.stores.LoadState
import com.wagerproof.app.features.agents.AgentColorPalette
import com.wagerproof.app.features.agents.agentSymbol

private val AccentGreen = Color(0xFF00E676)

/**
 * Native port of iOS `AgentLeaderboardView` (`components/agents/AgentLeaderboard.tsx`).
 * Renders the filter pill rows (when [showsFilters]), a ranked list of public
 * agents (top 3 wrapped in a glow halo), loading skeletons, and empty/error
 * states.
 *
 * Owns no state of its own — binds directly to the injected [store] (@Stable, so
 * its `by mutableStateOf` props are read directly and drive recomposition). Each
 * filter pill tap calls a `setX` setter whose guard re-runs the fetch.
 */
@OptIn(ExperimentalFoundationApi::class)
@Composable
fun AgentLeaderboard(
    store: LeaderboardStore,
    entitlements: AgentEntitlementsStore,
    showsFilters: Boolean = true,
    pinnedHeader: @Composable () -> Unit = {},
    onRowTap: (AgentLeaderboardEntry) -> Unit,
) {
    // First-load kickoff — mirrors iOS `.task { if idle { refresh() } }`.
    LaunchedEffect(Unit) {
        if (store.loadState is LoadState.Idle) store.refresh()
    }

    val listState = rememberLazyListState()
    val entries = store.entries
    val loadState = store.loadState

    LazyColumn(
        state = listState,
        modifier = Modifier.fillMaxWidth(),
    ) {
        // Pinned section header (the Agents hub lifts its tab/filter bar here).
        // The default empty header collapses to zero height.
        stickyHeader { pinnedHeader() }

        // Inline filter bar (hidden when the host lifts it into a glass header).
        if (showsFilters) {
            item(key = "filters") {
                LeaderboardFilterBar(
                    store = store,
                    modifier = Modifier.padding(start = 16.dp, end = 16.dp, top = 12.dp),
                )
            }
        }

        when (loadState) {
            // Refreshing = stale-while-revalidate: keep showing rows over the fetch.
            is LoadState.Idle, is LoadState.Loading, is LoadState.Refreshing -> {
                if (entries.isEmpty()) {
                    items(6) { SkeletonRow() }
                } else {
                    leaderboardRows(entries, store, entitlements, onRowTap)
                }
            }
            is LoadState.Loaded -> {
                if (entries.isEmpty()) {
                    item(key = "empty") { EmptyState() }
                } else {
                    leaderboardRows(entries, store, entitlements, onRowTap)
                }
            }
            is LoadState.Failed -> {
                item(key = "error") { ErrorState(loadState.message, store) }
            }
        }

        item(key = "bottom_spacer") { Spacer(Modifier.size(24.dp)) }
    }
}

/** Shared row emission used by both the loaded and stale-loading branches. */
private fun LazyListScope.leaderboardRows(
    entries: List<AgentLeaderboardEntry>,
    store: LeaderboardStore,
    entitlements: AgentEntitlementsStore,
    onRowTap: (AgentLeaderboardEntry) -> Unit,
) {
    // LazyListScope.items(count) member overload — keyed on the stable entry id.
    items(count = entries.size, key = { entries[it].id }) { idx ->
        val entry = entries[idx]
        LeaderboardRow(
            entry = entry,
            rank = idx + 1,
            lockStats = !entitlements.canViewAgentPicks,
            isEntitlementsLoading = entitlements.isLoading,
            isBottomMode = store.isBottomMode,
            onTap = { onRowTap(entry) },
            modifier = Modifier
                .padding(horizontal = 16.dp, vertical = 6.dp)
                // Cascade each row in as it replaces the loading shimmer.
                .staggeredAppear(index = idx),
        )
    }
}

// MARK: - Filter bar

/**
 * Sort-mode + timeframe pill rows plus the 10+ picks toggle. Extracted so it can
 * render inline here or be lifted into the Outliers tab's pinned glass header.
 * Each pill tap flips a store property whose setter re-runs the fetch.
 */
@Composable
private fun LeaderboardFilterBar(
    store: LeaderboardStore,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Row(
            modifier = Modifier
                .horizontalScroll(rememberScrollState())
                .padding(vertical = 2.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            AgentPerformanceService.LeaderboardSortMode.values().forEach { mode ->
                FilterPill(
                    label = mode.label,
                    isActive = store.sortMode == mode,
                ) { store.setSortMode(mode) }
            }
        }

        Row(
            modifier = Modifier
                .horizontalScroll(rememberScrollState())
                .padding(vertical = 2.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            AgentPerformanceService.LeaderboardTimeframe.values().forEach { tf ->
                FilterPill(
                    label = tf.label,
                    isActive = store.timeframe == tf,
                    isSubtle = true,
                ) { store.setTimeframe(tf) }
            }
            // Toggle-as-button — active state matches the pill treatment.
            FilterPill(
                label = "10+ picks",
                isActive = store.excludeUnder10Picks,
            ) { store.setExcludeUnder10Picks(!store.excludeUnder10Picks) }
        }
    }
}

@Composable
private fun FilterPill(
    label: String,
    isActive: Boolean,
    isSubtle: Boolean = false,
    onClick: () -> Unit,
) {
    val fill = if (isActive) {
        AccentGreen.copy(alpha = 0.15f)
    } else {
        AppColors.appBorder.copy(alpha = if (isSubtle) 0.3f else 0.5f)
    }
    val stroke = if (isActive) AccentGreen.copy(alpha = 0.45f) else AppColors.appBorder.copy(alpha = 0.3f)
    Box(
        Modifier
            .clip(CircleShape)
            .background(fill)
            .border(1.dp, stroke, CircleShape)
            .clickableNoRipple(onClick)
            .padding(horizontal = 12.dp, vertical = 6.dp),
    ) {
        Text(
            label,
            fontSize = 12.sp,
            fontWeight = FontWeight.Bold,
            color = if (isActive) AccentGreen else AppColors.appTextSecondary,
        )
    }
}

// MARK: - Row

@Composable
private fun LeaderboardRow(
    entry: AgentLeaderboardEntry,
    rank: Int,
    lockStats: Boolean,
    isEntitlementsLoading: Boolean,
    isBottomMode: Boolean,
    onTap: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(12.dp))
            .background(AppColors.appBorder.copy(alpha = 0.2f))
            .border(1.dp, AppColors.appBorder.copy(alpha = 0.3f), RoundedCornerShape(12.dp))
            .clickableNoRipple(onTap)
            .padding(12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        RankBadge(rank)
        AvatarSection(entry, rank, modifier = Modifier.weight(1f))
        StatsSection(entry, lockStats, isEntitlementsLoading)
        WinRateBadge(entry, isBottomMode)
        Icon(
            agentSymbol("chevron.right"),
            contentDescription = null,
            tint = AppColors.appTextSecondary,
            modifier = Modifier.size(13.dp),
        )
    }
}

@Composable
private fun RankBadge(rank: Int) {
    val color: Color
    val icon: String?
    when (rank) {
        1 -> { color = Color(0xFFFFD700); icon = "trophy.fill" }
        2 -> { color = Color(0xFFC0C0C0); icon = "medal.fill" }
        3 -> { color = Color(0xFFCD7F32); icon = "medal" }
        else -> { color = AppColors.appTextSecondary; icon = null }
    }
    Box(Modifier.size(width = 32.dp, height = 32.dp), contentAlignment = Alignment.Center) {
        if (icon != null) {
            Icon(
                agentSymbol(icon),
                contentDescription = null,
                tint = color,
                modifier = Modifier.size(22.dp),
            )
        } else {
            Text("$rank", fontSize = 16.sp, fontWeight = FontWeight.Bold, color = color)
        }
    }
}

@Composable
private fun AvatarSection(
    entry: AgentLeaderboardEntry,
    rank: Int,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        val avatarSize = if (rank <= 3) 44.dp else 36.dp
        val avatar: @Composable () -> Unit = {
            Box(
                Modifier
                    .size(avatarSize)
                    .clip(CircleShape)
                    .background(
                        Brush.linearGradient(AgentColorPalette.avatarGradient(entry.avatarColor)),
                    )
                    .padding(if (rank <= 3) 3.dp else 2.dp),
                contentAlignment = Alignment.Center,
            ) {
                PixelSpriteAvatar(spriteIndex = entry.spriteIndex, modifier = Modifier.fillMaxSize())
            }
        }
        // Top-3 rows get the animated glow halo.
        if (rank <= 3) {
            GlowingCardWrapper(color = AgentColorPalette.primary(entry.avatarColor), cornerRadius = 22.dp) {
                avatar()
            }
        } else {
            avatar()
        }

        Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(
                entry.name,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                color = AppColors.appTextPrimary,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                entry.preferredSports.take(2).forEach { sport ->
                    Text(
                        sport.label,
                        fontSize = 10.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = AppColors.appTextSecondary,
                    )
                }
                if (entry.preferredSports.size > 2) {
                    Text(
                        "+${entry.preferredSports.size - 2}",
                        fontSize = 10.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = AppColors.appTextSecondary,
                    )
                }
            }
        }
    }
}

@Composable
private fun StatsSection(
    entry: AgentLeaderboardEntry,
    lockStats: Boolean,
    isEntitlementsLoading: Boolean,
) {
    Column(
        modifier = Modifier.widthIn(min = 56.dp),
        horizontalAlignment = Alignment.End,
        verticalArrangement = Arrangement.spacedBy(2.dp),
    ) {
        Text(
            recordLabel(entry),
            fontSize = 12.sp,
            fontWeight = FontWeight.Bold,
            color = AppColors.appTextPrimary,
        )
        Box(contentAlignment = Alignment.Center) {
            if (isEntitlementsLoading) {
                CircularProgressIndicator(
                    modifier = Modifier.size(14.dp),
                    strokeWidth = 1.5.dp,
                    color = AppColors.appTextSecondary,
                )
            } else {
                Text(
                    netUnitsLabel(entry),
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Black,
                    color = if (entry.netUnits >= 0) AppColors.appWin else AppColors.appLoss,
                )
            }
            // Lock overlay when the viewer can't see agent picks.
            if (!isEntitlementsLoading && lockStats) {
                Box(
                    Modifier
                        .matchParentSize()
                        .clip(RoundedCornerShape(4.dp))
                        .background(AppColors.appSurface.copy(alpha = 0.85f)),
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(
                        agentSymbol("lock.fill"),
                        contentDescription = null,
                        tint = AppColors.appTextSecondary,
                        modifier = Modifier.size(9.dp),
                    )
                }
            }
        }
    }
}

@Composable
private fun WinRateBadge(entry: AgentLeaderboardEntry, isBottomMode: Boolean) {
    val label = entry.winRate?.let { String.format("%.1f%%", it * 100) } ?: "-"
    val color = when {
        isBottomMode -> {
            val wr = entry.winRate
            if (wr != null && wr < 0.35) AppColors.appLoss else Color(0xFFF97316)
        }
        else -> AccentGreen
    }
    Box(Modifier.size(width = 48.dp, height = 16.dp), contentAlignment = Alignment.Center) {
        Text(label, fontSize = 11.sp, fontWeight = FontWeight.Bold, color = color)
    }
}

private fun recordLabel(entry: AgentLeaderboardEntry): String {
    var s = "${entry.wins}-${entry.losses}"
    if (entry.pushes > 0) s += "-${entry.pushes}"
    return s
}

private fun netUnitsLabel(entry: AgentLeaderboardEntry): String {
    val sign = if (entry.netUnits >= 0) "+" else ""
    return String.format("%s%.2fu", sign, entry.netUnits)
}

// MARK: - Skeleton / empty / error

/**
 * Skeleton placeholder mirroring [LeaderboardRow]'s layout (rank block, avatar,
 * name + sport subtitle, trailing record/units + win-rate badge). The inner
 * group carries the shimmer sweep; the card chrome stays solid.
 */
@Composable
private fun SkeletonRow() {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 6.dp)
            .clip(RoundedCornerShape(12.dp))
            .background(AppColors.appBorder.copy(alpha = 0.2f))
            .padding(12.dp)
            .shimmering(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        SkeletonBlock(width = 28.dp, height = 20.dp, cornerRadius = 4.dp)
        SkeletonCircle(diameter = 36.dp)
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            SkeletonBlock(width = 120.dp, height = 14.dp)
            SkeletonBlock(width = 60.dp, height = 10.dp)
        }
        Column(
            horizontalAlignment = Alignment.End,
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            SkeletonBlock(width = 44.dp, height = 12.dp)
            SkeletonBlock(width = 52.dp, height = 14.dp)
        }
        SkeletonBlock(width = 40.dp, height = 11.dp)
    }
}

@Composable
private fun EmptyState() {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp)
            .clip(RoundedCornerShape(12.dp))
            .background(AppColors.appBorder.copy(alpha = 0.2f))
            .padding(vertical = 40.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Icon(
            agentSymbol("trophy"),
            contentDescription = null,
            tint = AppColors.appTextSecondary,
            modifier = Modifier.size(48.dp),
        )
        Text(
            "No public agents yet",
            fontSize = 16.sp,
            fontWeight = FontWeight.SemiBold,
            color = AppColors.appTextSecondary,
        )
        Text(
            "Be the first to make your agent public!",
            fontSize = 13.sp,
            color = AppColors.appTextSecondary,
        )
    }
}

@Composable
private fun ErrorState(message: String, store: LeaderboardStore) {
    val scope = androidx.compose.runtime.rememberCoroutineScope()
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Icon(
            agentSymbol("exclamationmark.triangle"),
            contentDescription = null,
            tint = AppColors.appTextSecondary,
            modifier = Modifier.size(40.dp),
        )
        Text(
            "Couldn't load leaderboard",
            fontSize = 16.sp,
            fontWeight = FontWeight.SemiBold,
            color = AppColors.appTextPrimary,
        )
        Text(message, fontSize = 13.sp, color = AppColors.appTextSecondary)
        Row(
            modifier = Modifier
                .clip(CircleShape)
                .background(AccentGreen)
                .clickableNoRipple { scope.launch { store.refresh() } }
                .padding(horizontal = 16.dp, vertical = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                agentSymbol("arrow.clockwise"),
                contentDescription = null,
                tint = Color.Black,
                modifier = Modifier.size(13.dp),
            )
            Text("Retry", fontSize = 13.sp, fontWeight = FontWeight.SemiBold, color = Color.Black)
        }
    }
}

// Rows/pills are whole-surface tappable with no ripple (iOS `.buttonStyle(.plain)`).
private fun Modifier.clickableNoRipple(onClick: () -> Unit): Modifier = composed {
    val interaction = remember { MutableInteractionSource() }
    clickable(interactionSource = interaction, indication = null, onClick = onClick)
}
